utils/test/resource_location_test.cpp

```diff
@@ -908,3 +908,37 @@ TEST(resource_location, replace_filename)
     EXPECT_EQ(resource_location::replace_filename(resource_location("schema://h/a/x/f?zzz"), "b").get_uri(),
         "schema://h/a/x/b?zzz");
 }
+
+TEST(resource_location, filename)
+{
+    EXPECT_EQ(resource_location("a").filename(), "a");
+    EXPECT_EQ(resource_location("x/a").filename(), "a");
+    EXPECT_EQ(resource_location("x/").filename(), "");
+    EXPECT_EQ(resource_location("schema:a").filename(), "a");
+    EXPECT_EQ(resource_location("schema://h/a").filename(), "a");
+    EXPECT_EQ(resource_location("schema://h/a/x/").filename(), "");
+    EXPECT_EQ(resource_location("schema://h/a").filename(), "a");
+    EXPECT_EQ(resource_location("schema://h/a/x/f?zzz").filename(), "f");
+}
+
+TEST(resource_location, parent)
+{
+    EXPECT_EQ(resource_location("a").parent().get_uri(), "");
+    EXPECT_EQ(resource_location("x/a").parent().get_uri(), "x");
+    EXPECT_EQ(resource_location("x/").parent().get_uri(), "x");
+    EXPECT_EQ(resource_location("schema:a").parent().get_uri(), "schema:");
+    EXPECT_EQ(resource_location("schema://h/a").parent().get_uri(), "schema://h");
+    EXPECT_EQ(resource_location("schema://h/a/x/").parent().get_uri(), "schema://h/a/x");
+    EXPECT_EQ(resource_location("schema://h/a").parent().get_uri(), "schema://h");
```

Comment:
If I remember correctly, the `/` in URLs represents a directory. If the slash is missing then the path is supposed to lead to a file. 

Therefore I suggest:
```suggestion
    EXPECT_EQ(resource_location("schema://h/a").parent().get_uri(), "schema://h/");
    EXPECT_EQ(resource_location("schema://h/a/x/").parent().get_uri(), "schema://h/a/x/");
    EXPECT_EQ(resource_location("schema://h/a").parent().get_uri(), "schema://h/");
```

Reply:
std::filesystem::path::parent_path() returns a path without the slash.
```
int main()
{
    std::cout << std::filesystem::path("/a/b/c/d").parent_path().string() << '\n';
    std::cout << std::filesystem::path("/a/b/c/").parent_path().string()  << '\n';
}
```
will print
```
/a/b/c
/a/b/c
```

Reply:
I agree that it's impractical but it would be RFC compliant. Let's discuss it in person